src/thor/route_action.cc

```diff
@@ -162,50 +262,69 @@ std::list<valhalla::odin::TripPath> thor_worker_t::path_arrive_by(
+      if (destination->path_edges().rbegin()->graph_id() == first_edge) {
         destination->mutable_path_edges()->SwapElements(0, destination->path_edges_size() - 1);
       }
       destination->mutable_path_edges()->RemoveLast();
     }
 
     // Get best path and keep it
     auto temp_path = get_path(path_algorithm, *origin, *destination, costing);
-    temp_path.swap(path);
+    first_edge = temp_path.front().edgeid;
+    temp_path.swap(path); // so we can append to path instead of prepend
 
     // Merge through legs by updating the time and splicing the lists
     if (!temp_path.empty()) {
       auto offset = path.back().elapsed_time;
       std::for_each(temp_path.begin(), temp_path.end(),
                     [offset](PathInfo& i) { i.elapsed_time += offset; });
+      // Connects via the same edge so we only need it once
       if (path.back().edgeid == temp_path.front().edgeid) {
         path.pop_back();
+      } else if (destination->type() == odin::Location::kVia) {
+        // Insert a route discontinuity if the paths meet at opposing edges and not
+        // at a graph node. Use path size - 1 as the index where the discontinuity lies.
+        via_discontinuity(*reader, *destination, path.back().edgeid, temp_path.front().edgeid, vias,
+                          temp_path.size(), true);
       }
       path.insert(path.end(), temp_path.begin(), temp_path.end());
     }
 
     // Build trip path for this leg and add to the result if this
     // location is a BREAK or if this is the last location
-    if (origin->type() == odin::Location::kBreak) {
+    if (origin->type() == odin::Location::kBreak || origin->type() == odin::Location::kBreakThrough) {
       // Move destination back to the last break and collect the throughs
       std::list<odin::Location> throughs;
-      while (destination->type() != odin::Location::kBreak) {
+      while (destination->type() != odin::Location::kBreak &&
+             destination->type() != odin::Location::kBreakThrough) {
         throughs.push_back(*destination);
         --destination;
       }
 
       // Create controller for default route attributes
       AttributesController controller;
 
+      // We have to flip the via indices because we built them in backwards order
+      decltype(vias) flipped;
+      flipped.reserve(vias.size());
+      for (const auto& kv : vias)
+        flipped.emplace(path.size() - kv.first, kv.second);
+      vias.swap(flipped);
+
       // Form output information based on path edges
       auto trip_path = thor::TripPathBuilder::Build(controller, *reader, mode_costing, path, *origin,
-                                                    *destination, throughs, interrupt);
+                                                    *destination, throughs, interrupt, &vias);
       path.clear();
+      vias.clear();
 
       // Keep the protobuf path
-      trip_paths.emplace_back(std::move(trip_path));
+      trip_paths.emplace_front(std::move(trip_path));
```

Comment:
this was a MASSIVE MASSIVE bug. i dont know how many folks are really using arrive_by with multi point routes, but this bug caused the legs of the route to be returned in reverse order! anyway, its fixed now :wink: